doc/monads/st.lean

```diff
@@ -0,0 +1,117 @@
+/-!
+# ST monad
+
+In the previous sections, [StateM](states.lean.md) and [ReaderM](readers.lean.md), you saw how
+`ReaderM` provides readonly state and `StateM` provides updatable state. There is another monad
+called `ST` which stands for "State Transfomer" which is a _stateful computation_: the computation
+is seen as transforming one state into another, where the new state will actually be constructed by
+modifying the old state in place.
+
+This is based on the work of John Launchbury and Simon L Peyton Jones in
+[Lazy Functional State Threads](https://www.microsoft.com/en-us/research/wp-content/uploads/1994/06/lazy-functional-state-threads.pdf).
+
+A state transfomer can be used as a first-class value which means it can be passed to a funciton,
+returned as a result and stored in a data structure, which means you can store it in a `ReaderM`
+context to make part of that context updatable.
```

Comment:
```suggestion
A reference can be used as a first-class value which means it can be passed to a function,
returned as a result and stored in a data structure, which means you can store it in a `ReaderM`
context to make the value being referenced updatable.
```